File: src/component-operations.mjs
import { emptyModel } from './model-state.mjs';
import { ensureMaterial } from './material-operations.mjs';
import { addMesh, addPrism, addCylinder } from './primitive-operations.mjs';
import { addBooleanCutout, addFaceWithHoles, addGableRoof, addPanelWithOpenings, addProfileExtrude, addShedRoof } from './profile-operations.mjs';
import { addAnalogStick, addBowedPanel, addDomedSurface, addFaceOnCylinder, addLoftBetweenProfiles, addLoftedSolid, addPipeBetweenPoints, addScrewHole, addShellFromFrontSideProfiles, addSweptPath } from './surface-operations.mjs';
import { addBeveledPanel, addBox, addButtonOnPanel, addChamfer, addEngravedLine, addFillet, addImagePlane, addRecess, addRib, addRoundedBox, addSlot, addSlotArray, addStandoffBoss, addText3d, addTextEmboss, addTextEngrave } from './product-operations.mjs';
import { addColumnGrid, addCurvedWall, addCurtainWall, addDoor, addFloorSlab, addFootprintSlab, addHipRoof, addParapetPath, addParkingStallArray, addPathSurface, addRailing, addRoofFootprint, addStairs, addTerrainMesh, addWall, addWallPath, addWindow } from './architecture-operations.mjs';
import { assertObjectIdentityAvailable, boxVertices, objectId, rotationZMatrix } from './object-identity.mjs';
import { applyTransform, normalizeQaMetadata, normalizeTransform, normalizeVector } from './operation-utils.mjs';
import { boundingBoxForVertices, mergeBoundingBoxes } from './snapshot.mjs';

export function addComponentDefinition(model, { name, size = [1000, 1000, 1000], material, operations }) {
  if (!name || typeof name !== 'string') throw new Error('component_definition operation requires a string name');

  if (operations !== undefined) {
    if (!Array.isArray(operations)) throw new Error(`${name}.operations must be an array`);
    const componentModel = emptyModel();
    for (const operation of operations) {
      applyComponentDefinitionOperation(componentModel, operation, name);
    }
    for (const [materialName, materialValue] of Object.entries(componentModel.materials)) {
      if (!model.materials[materialName]) model.materials[materialName] = materialValue;
    }
    const groups = componentModel.groups;
    const faces = groups.reduce((sum, group) => sum + group.faces, 0);
    const edges = groups.reduce((sum, group) => sum + group.edges, 0);
    model.component_definitions[name] = {
      name,
      faces,
      edges,
      groups,
      material: groups.find((group) => group.material)?.material || null,
      bounding_box: mergeBoundingBoxes(groups.map((group) => group.bounding_box))
    };
    return;
  }

  const normalizedSize = normalizeVector(size, [1000, 1000, 1000], `${name}.size`);
  if (normalizedSize.some((value) => value <= 0)) throw new Error(`${name}.size values must be positive`);
  const [w, d, h] = normalizedSize;
  model.component_definitions[name] = {
    name,
    faces: 6,
    edges: 12,
    groups: [],
    material: ensureMaterial(model, material),
    bounding_box: { min: [0, 0, 0], max: [w, d, h], w, d, h }
  };
}

function applyComponentDefinitionOperation(model, operation, componentName) {
  switch (operation.op) {
    case 'material':
      ensureMaterial(model, operation);
      break;
    case 'box':
      addBox(model, operation);
      break;
    case 'rounded_box':
      addRoundedBox(model, operation);
      break;
    case 'beveled_panel':
      addBeveledPanel(model, operation);
      break;
    case 'fillet':
      addFillet(model, operation);
      break;
    case 'chamfer':
      addChamfer(model, operation);
      break;
    case 'recess':
      addRecess(model, operation);
      break;
    case 'engraved_line':
      addEngravedLine(model, operation);
      break;
    case 'text_emboss':
      addTextEmboss(model, operation);
      break;
    case 'text_engrave':
      addTextEngrave(model, operation);
      break;
    case 'text_3d':
      addText3d(model, operation);
      break;
    case 'slot':
      addSlot(model, operation);
      break;
    case 'slot_array':
      addSlotArray(model, operation);
      break;
    case 'rib':
      addRib(model, operation);
      break;
    case 'standoff_boss':
      addStandoffBoss(model, operation);
      break;
    case 'button_on_panel':
      addButtonOnPanel(model, operation);
      break;
    case 'image_plane':
      addImagePlane(model, operation);
      break;
    case 'floor_slab':
      addFloorSlab(model, operation);
      break;
    case 'footprint_slab':
      addFootprintSlab(model, operation);
      break;
    case 'wall':
      addWall(model, operation);
      break;
    case 'wall_path':
      addWallPath(model, operation);
      break;
    case 'curved_wall':
      addCurvedWall(model, operation);
      break;
    case 'roof_footprint':
      addRoofFootprint(model, operation);
      break;
    case 'hip_roof':
      addHipRoof(model, operation);
      break;
    case 'parapet_path':
      addParapetPath(model, operation);
      break;
    case 'curtain_wall':
      addCurtainWall(model, operation);
      break;
    case 'column_grid':
      addColumnGrid(model, operation);
      break;
    case 'path_surface':
      addPathSurface(model, operation);
      break;
    case 'terrain_mesh':
      addTerrainMesh(model, operation);
      break;
    case 'parking_stall_array':
      addParkingStallArray(model, operation);
      break;
    case 'door':
      addDoor(model, operation);
      break;
    case 'window':
      addWindow(model, operation);
      break;
    case 'stairs':
      addStairs(model, operation);
      break;
    case 'railing':
      addRailing(model, operation);
      break;
    case 'panel_with_openings':
      addPanelWithOpenings(model, operation);
      break;
    case 'boolean_cutout':
      addBooleanCutout(model, operation);
      break;
    case 'mesh':
      addMesh(model, operation);
      break;
    case 'prism':
      addPrism(model, operation);
      break;
    case 'face_with_holes':
      addFaceWithHoles(model, operation);
      break;
    case 'profile_extrude':
      addProfileExtrude(model, operation);
      break;
    case 'gable_roof':
      addGableRoof(model, operation);
      break;
    case 'shed_roof':
      addShedRoof(model, operation);
      break;
    case 'cylinder':
      addCylinder(model, operation);
      break;
    case 'loft_between_profiles':
      addLoftBetweenProfiles(model, operation);
      break;
    case 'shell_from_front_side_profiles':
      addShellFromFrontSideProfiles(model, operation);
      break;
    case 'lofted_solid':
      addLoftedSolid(model, operation);
      break;
    case 'face_on_cylinder':
      addFaceOnCylinder(model, operation);
      break;
    case 'analog_stick':
      addAnalogStick(model, operation);
      break;
    case 'screw_hole':
      addScrewHole(model, operation);
      break;
    case 'pipe_between_points':
      addPipeBetweenPoints(model, operation);
      break;
    case 'swept_path':
      addSweptPath(model, operation);
      break;
    case 'domed_surface':
      addDomedSurface(model, operation);
      break;
    case 'bowed_panel':
      addBowedPanel(model, operation);
      break;
    default:
      throw new Error(`${componentName}.operations does not support op: ${operation.op}`);
  }
}

export function addComponentInstance(model, operation) {
  const { name, definition, origin = [0, 0, 0] } = operation;
  if (!name || typeof name !== 'string') throw new Error('component_instance operation requires a string name');
  const componentDefinition = model.component_definitions[definition];
  if (!componentDefinition) throw new Error(`${name}.definition not found: ${definition}`);
  const translation = normalizeVector(origin, [0, 0, 0], `${name}.origin`);
  const box = componentDefinition.bounding_box;
  const corners = boxVertices(box.min, [box.w, box.d, box.h]);
  const transform = operation.transform || {};
  const extraTranslate = normalizeVector(transform.translate ?? transform.translation ?? operation.translation ?? [0, 0, 0], [0, 0, 0], `${name}.transform.translate`);
  const normalizedTransform = normalizeTransform({ transform: { ...transform, translate: [translation[0] + extraTranslate[0], translation[1] + extraTranslate[1], translation[2] + extraTranslate[2]] } }, name);
  const transformedVertices = applyTransform(corners, {
    transform: {
      ...transform,
      translate: [translation[0] + extraTranslate[0], translation[1] + extraTranslate[1], translation[2] + extraTranslate[2]]
    }
  }, name);
  const boundingBox = boundingBoxForVertices(transformedVertices);
  const id = objectId(operation, name);
  assertObjectIdentityAvailable(model, { id, name });
  model.instances.push({
    id,
    name,
    definition,
    faces: componentDefinition.faces,
    edges: componentDefinition.edges,
    material: componentDefinition.material,
    transform: normalizedTransform,
    bounding_box: boundingBox,
    _vertices: transformedVertices,
    _orientation: rotationZMatrix(normalizedTransform.rotateZ || 0),
    qa: normalizeQaMetadata(operation.qa)
  });
}

File: sketchup_plugin/alma_sketchup_mcp/component_operations.rb
# frozen_string_literal: true

module AlmaSketchupMCP
  extend self

  def add_component_definition(model, operation)
    name = operation.fetch('name')
    definition = model.definitions[name] || model.definitions.add(name)
    definition.entities.clear!

    if operation.key?('operations')
      operations = operation['operations']
      raise "#{name}.operations must be an array" unless operations.is_a?(Array)

      operations.each { |child_operation| apply_component_definition_operation(definition.entities, child_operation, name) }
      return
    end

    size = vector(operation['size'] || [1000, 1000, 1000], "#{name}.size")
    material = operation['material']
    add_box(definition.entities, 'name' => "#{name}_Geometry", 'origin' => [0, 0, 0], 'size' => size, 'material' => material)
  end

  def apply_component_definition_operation(entities, operation, component_name)
    case operation['op']
    when 'material'
      ensure_material(operation)
    when 'box'
      add_box(entities, operation)
    when 'rounded_box'
      add_rounded_box(entities, operation)
    when 'beveled_panel'
      add_beveled_panel(entities, operation)
    when 'fillet'
      add_fillet(entities, operation)
    when 'chamfer'
      add_chamfer(entities, operation)
    when 'recess'
      add_recess(entities, operation)
    when 'engraved_line'
      add_engraved_line(entities, operation)
    when 'text_emboss'
      add_text_emboss(entities, operation)
    when 'text_engrave'
      add_text_engrave(entities, operation)
    when 'text_3d'
      add_text_3d(entities, operation)
    when 'slot'
      add_slot(entities, operation)
    when 'slot_array'
      add_slot_array(entities, operation)
    when 'rib'
      add_rib(entities, operation)
    when 'standoff_boss'
      add_standoff_boss(entities, operation)
    when 'button_on_panel'
      add_button_on_panel(entities, operation)
    when 'image_plane'
      add_image_plane(entities, operation)
    when 'floor_slab'
      add_floor_slab(entities, operation)
    when 'footprint_slab'
      add_footprint_slab(entities, operation)
    when 'wall'
      add_wall(entities, operation)
    when 'wall_path'
      add_wall_path(entities, operation)
    when 'curved_wall'
      add_curved_wall(entities, operation)
    when 'roof_footprint'
      add_roof_footprint(entities, operation)
    when 'hip_roof'
      add_hip_roof(entities, operation)
    when 'parapet_path'
      add_parapet_path(entities, operation)
    when 'curtain_wall'
      add_curtain_wall(entities, operation)
    when 'column_grid'
      add_column_grid(entities, operation)
    when 'path_surface'
      add_path_surface(entities, operation)
    when 'terrain_mesh'
      add_terrain_mesh(entities, operation)
    when 'parking_stall_array'
      add_parking_stall_array(entities, operation)
    when 'door'
      add_door(entities, operation)
    when 'window'
      add_window(entities, operation)
    when 'stairs'
      add_stairs(entities, operation)
    when 'railing'
      add_railing(entities, operation)
    when 'panel_with_openings'
      add_panel_with_openings(entities, operation)
    when 'boolean_cutout'
      add_boolean_cutout(entities, operation)
    when 'mesh'
      add_mesh(entities, operation)
    when 'geometry_input'
      add_geometry_input(entities, operation)
    when 'curve'
      add_curve(entities, operation)
    when 'arc_curve'
      add_arc_curve(entities, operation)
    when 'prism'
      add_prism(entities, operation)
    when 'face_with_holes'
      add_face_with_holes(entities, operation)
    when 'profile_extrude'
      add_profile_extrude(entities, operation)
    when 'gable_roof'
      add_gable_roof(entities, operation)
    when 'shed_roof'
      add_shed_roof(entities, operation)
    when 'cylinder'
      add_cylinder(entities, operation)
    when 'loft_between_profiles'
      add_loft_between_profiles(entities, operation)
    when 'shell_from_front_side_profiles'
      add_shell_from_front_side_profiles(entities, operation)
    when 'lofted_solid'
      add_lofted_solid(entities, operation)
    when 'face_on_cylinder'
      add_face_on_cylinder(entities, operation)
    when 'analog_stick'
      add_analog_stick(entities, operation)
    when 'screw_hole'
      add_screw_hole(entities, operation)
    when 'pipe_between_points'
      add_pipe_between_points(entities, operation)
    when 'swept_path'
      add_swept_path(entities, operation)
    when 'domed_surface'
      add_domed_surface(entities, operation)
    when 'bowed_panel'
      add_bowed_panel(entities, operation)
    when 'component_instance'
      add_component_instance_to_entities(entities, Sketchup.active_model, operation)
    else
      raise "#{component_name}.operations does not support op: #{operation['op']}"
    end
  end

  def add_component_instance(model, operation)
    add_component_instance_to_entities(model.entities, model, operation)
  end

  def add_component_instance_to_entities(entities, model, operation)
    name = operation.fetch('name')
    definition = model.definitions[operation.fetch('definition')]
    raise "#{name}.definition not found: #{operation['definition']}" unless definition

    origin = vector(operation['origin'] || [0, 0, 0], "#{name}.origin").map { |value| mm_to_model_units(value) }
    transform = operation['transform'] || {}
    rotate_z = transform['rotateZ'] || transform['rotationZ'] || operation['rotateZ']
    rotation = rotate_z ? Geom::Transformation.rotation(ORIGIN, Z_AXIS, rotate_z.to_f.degrees) : Geom::Transformation.new
    instance = entities.add_instance(definition, rotation)
    instance.name = name
    if instance.respond_to?(:set_attribute)
      id = object_id(operation, name)
      assert_entity_identity_available(instance, id, name)
      instance.set_attribute('AlmaSketchupMCP', 'id', id)
      qa = qa_metadata(operation)
      instance.set_attribute('AlmaSketchupMCP', 'qa', JSON.generate(qa)) if qa
    end
    translate = transform['translate'] || transform['translation'] || operation['translation'] || [0, 0, 0]
    extra_translate = vector(translate, "#{name}.transform.translate").map { |value| mm_to_model_units(value) }
    instance.transform!(Geom::Transformation.translation([
      origin[0] + extra_translate[0],
      origin[1] + extra_translate[1],
      origin[2] + extra_translate[2]
    ]))
    instance
  end

  def apply_transform(entity, operation)
    transform = operation['transform'] || {}
    translate = transform['translate'] || transform['translation'] || operation['translation']
    if translate
      vector_translate = vector(translate, "#{operation['name']}.transform.translate").map { |value| mm_to_model_units(value) }
      entity.transform!(Geom::Transformation.translation(vector_translate))
    end
    rotate_z = transform['rotateZ'] || transform['rotationZ'] || operation['rotateZ']
    return entity unless rotate_z

    angle = rotate_z.to_f.degrees
    entity.transform!(Geom::Transformation.rotation(ORIGIN, Z_AXIS, angle))
    entity
  end
end
